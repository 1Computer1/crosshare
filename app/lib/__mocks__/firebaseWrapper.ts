const firebaseTesting = jest.requireActual('@firebase/testing');
import type firebaseAdminType from 'firebase-admin';

export let App: firebase.app.App;
export let AdminApp: firebaseAdminType.app.App;

export const setApp = (app: firebase.app.App) => { App = app; };
export const setAdminApp = (app: firebaseAdminType.app.App) => { AdminApp = app; };

export const AuthProvider = null;

export const DeleteSentinal = firebaseTesting.firestore.FieldValue.delete();

export const ServerTimestamp = firebaseTesting.firestore.FieldValue.serverTimestamp();

export const TimestampClass = firebaseTesting.firestore.Timestamp;
export const AdminTimestamp = firebaseTesting.firestore.Timestamp;

let anonApp: firebase.app.App;
let anonUser: firebase.User;
export const setUpForSignInAnonymously = (app: firebase.app.App, user: firebase.User) => {
  anonApp = app;
  anonUser = user;
};

export const signInAnonymously = async () => {
  setApp(anonApp);
  return anonUser;
};
