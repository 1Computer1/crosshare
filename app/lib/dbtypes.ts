import * as t from 'io-ts';

// We import this instead of defining it here so that we can define it
// differently for the main react app and the 'functions' subdirectory
import { timestamp } from './timestamp';
export { timestamp } from './timestamp';

const CommentV = t.type({
  /** comment text */
  c: t.string,
  /** author id */
  a: t.string,
  /** author display name */
  n: t.string,
  /** author solve time in fractional seconds */
  t: t.number,
  /** author did cheat? */
  ch: t.boolean,
  /** comment publish timestamp */
  p: timestamp
});
type CommentT = t.TypeOf<typeof CommentV>;

export interface CommentWithRepliesT extends CommentT {
  /** comment id */
  i: string,
  /** replies */
  r?: Array<CommentWithRepliesT>
}

const CommentWithRepliesV: t.Type<CommentWithRepliesT> = t.recursion('CommentWithReplies', () =>
  t.intersection([
    CommentV,
    t.type({
      /** comment id */
      i: t.string,
    }),
    t.partial({
      /** replies */
      r: t.array(CommentWithRepliesV)
    })
  ])
);

export const CommentForModerationV = t.intersection([
  CommentV,
  t.type({
    /** puzzle id */
    pid: t.string,
    /** id of the comment this is a reply to */
    rt: t.union([t.string, t.null])
  })
]);
export type CommentForModerationT = t.TypeOf<typeof CommentForModerationV>;

export const CommentForModerationWithIdV = t.intersection([
  CommentForModerationV,
  t.type({
    /** comment id */
    i: t.string,
  })
]);
export type CommentForModerationWithIdT = t.TypeOf<typeof CommentForModerationWithIdV>;

const DBPuzzleMandatoryV = t.type({
  /** created at */
  ca: timestamp,
  /** author's user id */
  a: t.string,
  /** author's display name */
  n: t.string,
  /** category */
  c: t.union([t.string, t.null]),
  /** is this puzzle moderated? */
  m: t.boolean,
  /** timestamp when the puzzle goes live */
  p: t.union([timestamp, t.null]),
  /** title */
  t: t.string,
  /** grid width / columns */
  w: t.number,
  /** grid height / rows */
  h: t.number,
  /** across clue strings */
  ac: t.array(t.string),
  /** across clue display numbers */
  an: t.array(t.number),
  /** down clue strings */
  dc: t.array(t.string),
  /** down clue display numbers */
  dn: t.array(t.number),
  /** grid (solution) */
  g: t.array(t.string),
});
const DBPuzzleOptionalV = t.partial({
  /** highlighted cell indexes */
  hs: t.array(t.number),
  /** use shade instead of circle for highlight? */
  s: t.boolean,
  /** comments */
  cs: t.array(CommentWithRepliesV),
  /** is this puzzle featured (should appear on homepage) */
  f: t.boolean,
});
export const DBPuzzleV = t.intersection([DBPuzzleMandatoryV, DBPuzzleOptionalV]);
export type DBPuzzleT = t.TypeOf<typeof DBPuzzleV>;

export const TimestampedPuzzleV = downloadTimestamped(DBPuzzleV);
export type TimestampedPuzzleT = t.TypeOf<typeof TimestampedPuzzleV>;

const PlayBaseV = t.type({
  /** crossword id */
  c: t.string,
  /** updated at */
  ua: timestamp,
  /** filled in grid */
  g: t.array(t.string),
  /** play time of last update to each cell, in fractional seconds */
  ct: t.array(t.number),
  /** update iteration count per cell */
  uc: t.array(t.number),
  /** list of (checked) correct cells */
  vc: t.array(t.number),
  /** list of (checked) wrong cells */
  wc: t.array(t.number),
  /** list of cells ever marked wrong */
  we: t.array(t.number),
  /** list of revealed cells */
  rc: t.array(t.number),
  /** total play time, in fractional seconds */
  t: t.number,
  /** did "cheat"? */
  ch: t.boolean,
  /** finished the puzzle? */
  f: t.boolean,
});

export const LegacyPlayV = t.intersection([
  PlayBaseV,
  t.type({
    /** user id */
    u: t.string,
  }),
  t.partial({
    /** puzzle title, optional for legacy plays in the db */
    n: t.string,
  })
]);
export type LegacyPlayT = t.TypeOf<typeof LegacyPlayV>;

export const PlayV = t.intersection([
  PlayBaseV,
  t.type({
    /** user id */
    u: t.string,
    /** puzzle title */
    n: t.string,
  })
]);
export type PlayT = t.TypeOf<typeof PlayV>;

// We don't need a user id in local storage, but we do need a title
export const PlayWithoutUserV = t.intersection([
  PlayBaseV,
  t.type({
    /** puzzle title */
    n: t.string,
  })
]);
export type PlayWithoutUserT = t.TypeOf<typeof PlayWithoutUserV>;

export function downloadTimestamped<A>(type: t.Type<A>) {
  return t.type({
    downloadedAt: timestamp,
    data: type,
  });
}

export function downloadOptionallyTimestamped<A>(type: t.Type<A>) {
  return t.type({
    downloadedAt: t.union([timestamp, t.null]),
    data: type,
  });
}

export function getDateString(pd: Date) {
  return pd.getUTCFullYear() + '-' + pd.getUTCMonth() + '-' + pd.getUTCDate();
}

export function prettifyDateString(dateString: string) {
  const groups = dateString.match(/^(\d+)-(\d+)-(\d+)$/);
  if (!groups) {
    throw new Error('Bad date string: ' + dateString);
  }
  return (parseInt(groups[2]) + 1) + '/' + parseInt(groups[3]) + '/' + groups[1];
}

export const PuzzleStatsV = t.type({
  /** author id, denormalized for security rules purposes. */
  a: t.string,
  /** updated at */
  ua: timestamp,
  /** total completions */
  n: t.number,
  /** total completions without cheats */
  s: t.number,
  /** total time spent on puzzle, in fractional seconds */
  nt: t.number,
  /** total time spent by those w/o cheats */
  st: t.number,
  /** total play time of last update to each cell, in fractional seconds */
  ct: t.array(t.number),
  /** total update iteration count for each cell */
  uc: t.array(t.number),
});
export type PuzzleStatsT = t.TypeOf<typeof PuzzleStatsV>;

export const DailyStatsV = t.type({
  /** updated at */
  ua: timestamp,
  /** total completions */
  n: t.number,
  /** user ids with completions */
  u: t.array(t.string),
  /** completions by puzzleId */
  c: t.record(t.string, t.number),
  /** completions by hour (as UTC 0-23) */
  h: t.array(t.number),
});
export type DailyStatsT = t.TypeOf<typeof DailyStatsV>;

export const CronStatusV = t.type({
  ranAt: timestamp,
});
export type CronStatusT = t.TypeOf<typeof CronStatusV>;

/** created at, title */
const AuthoredPuzzleV = t.tuple([timestamp, t.string]);
export type AuthoredPuzzleT = t.TypeOf<typeof AuthoredPuzzleV>;

/** keys are puzzle ids */
export const AuthoredPuzzlesV = t.record(t.string, AuthoredPuzzleV);

/** date string -> puzzle id */
export const CategoryIndexV = t.record(t.string, t.string);
export type CategoryIndexT = t.TypeOf<typeof CategoryIndexV>;
