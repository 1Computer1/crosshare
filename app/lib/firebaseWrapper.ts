declare var firebase: typeof import('firebase');

// On the server we need to require firebase - on client it's done w/ script tags
if (typeof window === "undefined") {
  globalThis.firebase = require("firebase");
  require("firebase/firestore");
}

const firebaseConfig = {
  apiKey: "AIzaSyBrmmBf91peVT5T_Z7N3z9oizsPH5u2pUc",
  authDomain: "auth.crosshare.org",
  databaseURL: "https://mdcrosshare.firebaseio.com",
  projectId: "mdcrosshare",
  storageBucket: "mdcrosshare.appspot.com",
  messagingSenderId: "603173482014",
  appId: "1:603173482014:web:98d7d820731b7c5eaa080f",
  measurementId: "G-LTLN7Z4XBS"
};

// Initialize Firebase
export let App: firebase.app.App;
if (firebase.apps.length) {
  App = firebase.apps[0];
} else {
  App = firebase.initializeApp(firebaseConfig);
  if (typeof window !== "undefined") {
    if (process.env.NODE_ENV === 'production') {
      App.performance();
    }
  }
  if (process.env.NEXT_PUBLIC_USE_FIREBASE_EMULATOR) {
    var db = App.firestore();
    db.settings({
      host: "localhost:8080",
      ssl: false
    });
  }
}

export const setApp = (app: firebase.app.App) => { App = app };

export const AuthProvider = new firebase.auth.GoogleAuthProvider();

export const DeleteSentinal = firebase.firestore.FieldValue.delete();

export const TimestampClass = firebase.firestore.Timestamp;
export type TimestampType = firebase.firestore.Timestamp;
