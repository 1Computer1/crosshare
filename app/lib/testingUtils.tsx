import { ReactElement, ComponentType } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { matchers, createSerializer } from 'jest-emotion';

import { AuthContext } from '../components/AuthContext';

import type firebaseTypes from 'firebase';

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers);
expect.addSnapshotSerializer(createSerializer());

export const anonymousUser = {
  uid: 'anonymous-user-id',
  isAnonymous: true
} as firebaseTypes.User;

const WithAllProviders: ComponentType = ({ children }) => {
  return (
    <AuthContext.Provider value={{ user: anonymousUser, isAdmin: false, loadingUser: false, error: undefined }}>
      {children}
    </AuthContext.Provider>
  )
}

function wrappedRender(
  ui: ReactElement,
  options?: RenderOptions
): RenderResult {

  return render(ui, {
    wrapper: WithAllProviders,
    ...options,
  });
}
// re-export everything
export * from '@testing-library/react';

// override render method
export { wrappedRender as render };
