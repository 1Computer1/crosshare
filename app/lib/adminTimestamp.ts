/* this is the firebase timestamp type for the main react project! */

import * as t from 'io-ts';
import { either } from 'fp-ts/lib/Either';

import { AdminTimestamp } from './firebaseWrapper';
import type firebaseAdminType from 'firebase-admin';

const isFirestoreTimestamp = (u: unknown): u is firebaseAdminType.firestore.Timestamp =>
  u ? u instanceof AdminTimestamp : false;

const validateTimestamp: t.Validate<unknown, firebaseAdminType.firestore.Timestamp> = (i, c) => {
  if (isFirestoreTimestamp(i)) {
    return t.success(i);
  }
  return either.chain(
    t.type({ seconds: t.number, nanoseconds: t.number }).validate(i, c),
    obj => t.success(new AdminTimestamp(obj.seconds, obj.nanoseconds))
  );
};

export const adminTimestamp = new t.Type<firebaseAdminType.firestore.Timestamp>(
  'AdminTimestamp',
  isFirestoreTimestamp,
  validateTimestamp,
  t.identity,
);
