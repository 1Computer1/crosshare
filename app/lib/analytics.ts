/* This gets used by the analytics cron function in the `functions` directory.

It lives here so we can test it. */
import type firebase from 'firebase-admin';
import { PathReporter } from 'io-ts/lib/PathReporter';
import { isRight } from 'fp-ts/lib/Either';

import {
  LegacyPlayV,
  PuzzleStatsT,
  PuzzleStatsV,
  DailyStatsV,
  DailyStatsT,
  DBPuzzleV,
  DBPuzzleT,
  getDateString,
} from './dbtypes';

import { TimestampType } from './firebaseWrapper';

export async function runAnalytics(
  db: firebase.firestore.Firestore,
  startTimestamp: TimestampType,
  endTimestamp: TimestampType
) {
  console.log(
    'Updating analytics btwn',
    startTimestamp.toDate().toLocaleString(),
    endTimestamp.toDate().toLocaleString()
  );
  const value = await db
    .collection('p')
    .where('f', '==', true)
    .where('ua', '>=', startTimestamp)
    .where('ua', '<', endTimestamp)
    .orderBy('ua', 'asc')
    .get();
  console.log('Updating analytics for ' + value.size + ' plays');

  const puzzleMap: Map<string, DBPuzzleT> = new Map();
  const puzzleStatsMap: Map<string, PuzzleStatsT> = new Map();
  const dailyStatsMap: Map<string, DailyStatsT> = new Map();

  // Get puzzle obj from cache or db
  async function getPuzzle(puzzleId: string): Promise<DBPuzzleT> {
    const puzzle = puzzleMap.get(puzzleId);
    if (puzzle) {
      return puzzle;
    }
    const puzzleRes = await db.collection('c').doc(puzzleId).get();
    if (!puzzleRes.exists) {
      throw new Error('Missing puzzle but have play: ' + puzzleId);
    }
    const dbpuzzle = DBPuzzleV.decode(puzzleRes.data());
    if (!isRight(dbpuzzle)) {
      console.error(PathReporter.report(dbpuzzle).join(','));
      throw new Error('Malformed puzzle');
    }
    puzzleMap.set(puzzleId, dbpuzzle.right);
    return dbpuzzle.right;
  }

  for (const doc of value.docs) {
    const validationResult = LegacyPlayV.decode(doc.data());
    if (!isRight(validationResult)) {
      console.error(PathReporter.report(validationResult).join(','));
      throw new Error('Malformed play');
    }
    const play = validationResult.right;

    // Make sure we have a puzzle stats obj in the cache for the relevant puzzle
    let puzzleStats = puzzleStatsMap.get(play.c);
    if (!puzzleStats) {
      // get puzzle stats from db or create a new stats object
      const psvalue = await db.collection('s').doc(play.c).get();
      if (psvalue.exists) {
        const result = PuzzleStatsV.decode(psvalue.data());
        if (!isRight(result)) {
          console.error(PathReporter.report(result).join(','));
          throw new Error('Malformed puzzle stats');
        }
        puzzleStats = result.right;
        puzzleStatsMap.set(play.c, puzzleStats);
      } else {
        const puzzle = await getPuzzle(play.c);
        puzzleStats = {
          a: puzzle.a,
          ua: endTimestamp,
          n: 0,
          s: 0,
          nt: 0,
          st: 0,
          ct: [],
          uc: [],
          sct: Array.from(
            { length: 6 },
            () => Math.random().toString(36)[2]
          ).join(''),
        };
        puzzleStatsMap.set(play.c, puzzleStats);
      }
    }

    // don't count the author's play in puzzle stats
    if (puzzleStats.a === play.u) {
      continue;
    }

    puzzleStats.ua = endTimestamp;
    puzzleStats.n += 1;
    puzzleStats.nt += play.t;
    if (!play.ch) {
      puzzleStats.s += 1;
      puzzleStats.st += play.t;
    }
    const maxTime = Math.max(...play.ct);
    const maxCount = Math.max(...play.uc);
    for (let i = 0; i < play.ct.length; i += 1) {
      let updateTime = play.ct[i];
      const updateIters = play.uc[i];
      if (play.rc.indexOf(i) !== -1 || play.we.indexOf(i) !== -1) {
        /* If a cell was revealed or checked & wrong, make it's update time the
         * end of the play. This way cheat cells always show as taking the
         * longest for the user in question. */
        updateTime = play.t;
      }
      puzzleStats.ct[i] = (puzzleStats.ct[i] || 0) + updateTime / maxTime;
      puzzleStats.uc[i] = (puzzleStats.uc[i] || 0) + updateIters / maxCount;
    }

    // Next update daily stats for the relevant date
    const pd = play.ua.toDate();
    const dateString = getDateString(pd);
    let dailyStats = dailyStatsMap.get(dateString);
    if (!dailyStats) {
      // get daily stats from db or create a new stats object
      const dsvalue = await db.collection('ds').doc(dateString).get();
      if (dsvalue.exists) {
        const result = DailyStatsV.decode(dsvalue.data());
        if (!isRight(result)) {
          console.error(PathReporter.report(result).join(','));
          throw new Error('Malformed daily stats');
        }
        dailyStats = result.right;
        dailyStatsMap.set(dateString, dailyStats);
      } else {
        dailyStats = {
          ua: endTimestamp,
          n: 0,
          u: [],
          c: {},
          i: {},
          h: Array<number>(24).fill(0),
        };
        dailyStatsMap.set(dateString, dailyStats);
      }
    }
    dailyStats.ua = endTimestamp;
    dailyStats.n += 1;
    if (dailyStats.u.indexOf(play.u) === -1) {
      dailyStats.u.push(play.u);
    }
    dailyStats.c[play.c] = (dailyStats.c[play.c] || 0) + 1;
    if (dailyStats.i && !dailyStats.i[play.c]) {
      const puzzle = await getPuzzle(play.c);
      dailyStats.i[play.c] = [puzzle.t, puzzle.n, puzzle.a];
    }
    const hour = pd.getUTCHours();
    dailyStats.h[hour] = (dailyStats.h[hour] || 0) + 1;
  }

  console.log('Done, writing ' + puzzleStatsMap.size + ' puzzle stats objects');
  for (const [crosswordId, puzzleStats] of puzzleStatsMap.entries()) {
    await db.collection('s').doc(crosswordId).set(puzzleStats);
  }
  console.log('Writing ' + dailyStatsMap.size + ' daily stats objects');
  for (const [dateString, dailyStats] of dailyStatsMap.entries()) {
    await db.collection('ds').doc(dateString).set(dailyStats);
  }
}
