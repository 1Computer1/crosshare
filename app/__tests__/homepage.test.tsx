import { render } from '../lib/testingUtils';
import { setApp, TimestampClass } from '../lib/firebaseWrapper';
import * as firebaseTesting from '@firebase/testing';
import { DBPuzzleT, getDateString } from '../lib/dbtypes';
import HomePage, { getServerSideProps } from '../pages/index';

let adminApp: firebase.app.App;
let app: firebase.app.App;

let dailymini: DBPuzzleT;
const dailyminiId = 'dailymini';
let featuredPuzzle: DBPuzzleT;
const featuredId = 'published-puzzle';
let nonFeaturedPuzzle: DBPuzzleT;
const nonFeaturedId = 'pending-puzzle';

jest.mock('../lib/firebaseWrapper');

const TEST_DB = 'homepagetest';

beforeAll(async () => {
  adminApp = firebaseTesting.initializeAdminApp({ projectId: TEST_DB }) as firebase.app.App;
  app = firebaseTesting.initializeTestApp({ projectId: TEST_DB }) as firebase.app.App;

  const hourAgo = new Date();
  hourAgo.setMinutes(hourAgo.getMinutes() - 60);
  const hourFromNow = new Date();
  hourFromNow.setMinutes(hourFromNow.getMinutes() + 60);

  dailymini = {
    c: 'dailymini',
    m: true,
    t: 'Raises, as young',
    dn: [1, 2, 3, 4, 5],
    ac:
      [' Cobbler\'s forms',
        'Absolutely perfect',
        'Spike Lee\'s "She\'s ___ Have It"',
        'English class assignment',
        'Raises, as young'],
    dc:
      ['Hybrid whose father is a lion',
        '___ of reality (wake-up call)',
        '___ date (makes wedding plans)',
        'Middle Ages invader',
        'Has a great night at the comedy club'],
    p: TimestampClass.fromDate(hourAgo),
    a: 'puzzleauthor',
    an: [1, 6, 7, 8, 9],
    g:
      ['L',
        'A',
        'S',
        'T',
        'S',
        'I',
        'D',
        'E',
        'A',
        'L',
        'G',
        'O',
        'T',
        'T',
        'A',
        'E',
        'S',
        'S',
        'A',
        'Y',
        'R',
        'E',
        'A',
        'R',
        'S'],
    h: 5,
    w: 5,
    cs:
      [{
        c:
          'A couple of two-worders today which I don\'t love, but I hope you all got it anyway!',
        i: 'LwgoVx0BAskM4wVJyoLj',
        t: 36.009,
        p: TimestampClass.now(),
        a: 'fSEwJorvqOMK5UhNMHa4mu48izl1',
        n: 'Mike D',
        ch: false,
      }],
    n: 'Mike D'
  };
  await adminApp.firestore().collection('c').doc(dailyminiId).set(dailymini);
  await adminApp.firestore().collection('categories').doc('dailymini').set({ [getDateString(new Date())]: dailyminiId });

  featuredPuzzle = {
    c: null,
    m: true,
    f: true,
    t: 'Raises, as young',
    dn: [1, 2, 3, 4, 5],
    ac:
      [' Cobbler\'s forms',
        'Absolutely perfect',
        'Spike Lee\'s "She\'s ___ Have It"',
        'English class assignment',
        'Raises, as young'],
    dc:
      ['Hybrid whose father is a lion',
        '___ of reality (wake-up call)',
        '___ date (makes wedding plans)',
        'Middle Ages invader',
        'Has a great night at the comedy club'],
    p: TimestampClass.fromDate(hourAgo),
    a: 'puzzleauthor',
    an: [1, 6, 7, 8, 9],
    g:
      ['L',
        'A',
        'S',
        'T',
        'S',
        'I',
        'D',
        'E',
        'A',
        'L',
        'G',
        'O',
        'T',
        'T',
        'A',
        'E',
        'S',
        'S',
        'A',
        'Y',
        'R',
        'E',
        'A',
        'R',
        'S'],
    h: 5,
    w: 5,
    cs:
      [{
        c:
          'A couple of two-worders today which I don\'t love, but I hope you all got it anyway!',
        i: 'LwgoVx0BAskM4wVJyoLj',
        t: 36.009,
        p: TimestampClass.now(),
        a: 'fSEwJorvqOMK5UhNMHa4mu48izl1',
        n: 'Mike D',
        ch: false,
      }],
    n: 'Mike D'
  };
  await adminApp.firestore().collection('c').doc(featuredId).set(featuredPuzzle);

  nonFeaturedPuzzle = {
    c: null,
    m: true,
    t: 'Raises, as young',
    dn: [1, 2, 3, 4, 5],
    ac:
      [' Cobbler\'s forms',
        'Absolutely perfect',
        'Spike Lee\'s "She\'s ___ Have It"',
        'English class assignment',
        'Raises, as young'],
    dc:
      ['Hybrid whose father is a lion',
        '___ of reality (wake-up call)',
        '___ date (makes wedding plans)',
        'Middle Ages invader',
        'Has a great night at the comedy club'],
    p: TimestampClass.fromDate(hourAgo),
    a: 'puzzleauthor',
    an: [1, 6, 7, 8, 9],
    g:
      ['L',
        'A',
        'S',
        'T',
        'S',
        'I',
        'D',
        'E',
        'A',
        'L',
        'G',
        'O',
        'T',
        'T',
        'A',
        'E',
        'S',
        'S',
        'A',
        'Y',
        'R',
        'E',
        'A',
        'R',
        'S'],
    h: 5,
    w: 5,
    cs:
      [{
        c:
          'A couple of two-worders today which I don\'t love, but I hope you all got it anyway!',
        i: 'LwgoVx0BAskM4wVJyoLj',
        t: 36.009,
        p: TimestampClass.now(),
        a: 'fSEwJorvqOMK5UhNMHa4mu48izl1',
        n: 'Mike D',
        ch: false,
      }],
    n: 'Mike D'
  };
  await adminApp.firestore().collection('c').doc(nonFeaturedId).set(nonFeaturedPuzzle);
});

afterAll(async () => {
  await adminApp.delete();
  await app.delete();
});

test('get home page', async () => {
  setApp(app);

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const props = await getServerSideProps({ res: { setHeader: jest.fn() } } as any);
  const r = render(<HomePage {...props.props} />, {});
  expect(r.container).toMatchSnapshot();
});
