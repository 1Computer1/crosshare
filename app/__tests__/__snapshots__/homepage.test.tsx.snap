// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get home page 1`] = `
.emotion-6 {
  height: 35px;
  background-color: var(--primary);
}

.emotion-5 {
  padding: 0 10px;
  height: 100%;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 31px;
}

.emotion-1 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  cursor: pointer;
}

.emotion-0 {
  margin-left: 5px;
  display: none;
  color: var(--text);
  font-size: 25px;
}

@media (min-width:576px) {
  .emotion-0 {
    display: inline-block;
  }
}

.emotion-4 {
  background-color: transparent;
  border: none;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline;
  margin: 0;
  padding: 0 0.45em;
  color: var(--text);
}

.emotion-4:hover,
.emotion-4:focus {
  color: var(--text);
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: var(--top-bar-hover);
}

.emotion-2 {
  vertical-align: baseline;
  font-size: 25px;
}

.emotion-3 {
  margin-left: 5px;
  vertical-align: middle;
  display: none;
  font-size: 15px;
}

@media (min-width:576px) {
  .emotion-3 {
    display: inline-block;
  }
}

.emotion-17 {
  margin: 1em;
}

.emotion-7 {
  margin-bottom: 1em;
}

.emotion-11 {
  overflow: auto;
  display: inline-block;
  width: 100%;
}

@media (min-width:576px) {
  .emotion-11 {
    width: 50%;
  }
}

.emotion-10 {
  color: var(--link);
}

.emotion-10:hover {
  color: var(--link-hover);
}

.emotion-9 {
  position: relative;
  vertical-align: top !important;
  float: left;
  font-size: 4em;
  margin-right: 0.3em;
}

.emotion-8 {
  vertical-align: top;
}

.emotion-16 {
  margin-top: 1em;
}

<div>
  <header
    class="emotion-6"
  >
    <div
      class="emotion-5"
    >
      <a
        class="emotion-1"
        href="/"
        title="Crosshare Home"
      >
        <svg
          height="31"
          viewBox="0 0 16 16"
          width="31"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="var(--logo-white)"
          >
            <path
              d="M1 10h14v5H1z"
            />
            <path
              d="M3 9h10v1H3zm0-7h5v3H3z"
            />
            <path
              d="M6 5h1v5H6zm4-2h2v6h-2z"
            />
            <path
              d="M12 1h2v3h-2z"
            />
            <path
              d="M14 1h1v1h-1z"
            />
          </g>
          <path
            d="M13 0h3v1h-3zM4 1h3v1H4z"
          />
          <path
            d="M12 1h1v1h-1zm3 0h1v1h-1zM3 2h1v1H3zm3 0h2v1H6z"
          />
          <path
            d="M11 2h1v1h-1zm3 0h1v1h-1zM2 3h1v3H2zm3 0h1v6H5z"
          />
          <path
            d="M7 3h1v7H7zm3 0h1v1h-1zm3 0h1v1h-1zM4 4h1v1H4z"
          />
          <path
            d="M9 4h1v5H9zm3 0h1v1h-1z"
          />
          <path
            d="M3 5h1v1H3zm8 0h1v4h-1zM4 8h1v1H4z"
          />
          <path
            d="M3 9h1v1H3zm5 0h1v1H8z"
          />
          <path
            d="M12 9h1v1h-1zM1 10h2v1H1z"
          />
          <path
            d="M5 10h1v1H5zm8 0h2v1h-2z"
          />
          <path
            d="M0 11h1v3H0zm15 0h1v3h-1zm-9 0h1v1H6z"
          />
          <path
            d="M4 12h2v1H4z"
          />
          <path
            d="M5 13h2v1H5zm4 0h2v1H9zm-8 1h3v1H1zm11 0h3v1h-3z"
          />
          <path
            d="M4 15h8v1H4z"
          />
          <rect
            height="1"
            width="1"
            x="10"
            y="10"
          >
            <animate
              attributeName="width"
              begin="10;a.end+10"
              calcMode="discrete"
              dur="0.5s"
              id="a"
              values="1;0"
            />
          </rect>
          <rect
            height="1"
            width="1"
            x="9"
            y="11"
          >
            <animate
              attributeName="width"
              begin="10;b.end+10"
              calcMode="discrete"
              dur="0.5s"
              id="b"
              values="1;3"
            />
          </rect>
          <rect
            height="1"
            width="2"
            x="10"
            y="12"
          >
            <animate
              attributeName="width"
              begin="10;c.end+10"
              calcMode="discrete"
              dur="0.5s"
              id="c"
              values="2;0"
            />
          </rect>
        </svg>
        <span
          class="emotion-0"
        >
          CROSSHARE
        </span>
      </a>
      
      <a
        class="emotion-4"
        href="/account"
        title="Account"
      >
        <span
          class="emotion-2"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
            />
          </svg>
        </span>
        <span
          class="emotion-3"
        >
          Account
        </span>
      </a>
    </div>
  </header>
  <div
    class="emotion-17"
  >
    <p
      class="emotion-7"
    >
      Crosshare is the best place to create, share and solve crossword puzzles.
    </p>
    <h2>
      Daily Mini
    </h2>
    <div
      class="emotion-11"
    >
      <a
        class="emotion-10"
        href="/crosswords/dailymini"
      >
        <div
          class="emotion-9"
        >
          <svg
            class="emotion-8"
            fill="currentColor"
            height="1em"
            viewBox="0 0 100 100"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="currentColor"
              height="45"
              width="45"
              x="5"
              y="50"
            />
            <rect
              fill="transparent"
              height="90"
              rx="5"
              ry="5"
              stroke="currentColor"
              stroke-width="5"
              width="90"
              x="5"
              y="5"
            />
            
            
            
          </svg>
          
        </div>
        <h3>
          Today's daily mini crossword
           
          
        </h3>
      </a>
      <p>
        by 
        Mike D
      </p>
      <p>
        <a
          href="/categories/dailymini"
        >
          Play previous daily minis
        </a>
      </p>
    </div>
    <h2>
      Share a Puzzle
    </h2>
    <p>
      <a
        href="/upload"
      >
        Upload a .puz to get a Crosshare link to share with solvers
      </a>
    </p>
    <h2>
      Featured Puzzles
    </h2>
    <div
      class="emotion-11"
    >
      <a
        class="emotion-10"
        href="/crosswords/published-puzzle"
      >
        <div
          class="emotion-9"
        >
          <svg
            class="emotion-8"
            fill="currentColor"
            height="1em"
            viewBox="0 0 100 100"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="currentColor"
              height="45"
              width="45"
              x="5"
              y="50"
            />
            <rect
              fill="transparent"
              height="90"
              rx="5"
              ry="5"
              stroke="currentColor"
              stroke-width="5"
              width="90"
              x="5"
              y="5"
            />
            <text
              font-size="45"
              lengthAdjust="spacingAndGlyphs"
              text-anchor="middle"
              textLength="40"
              x="28"
              y="45"
            >
              5
            </text>
            <text
              font-size="45"
              text-anchor="middle"
              textLength="40"
              x="72"
              y="45"
            >
              x
            </text>
            <text
              font-size="45"
              lengthAdjust="spacingAndGlyphs"
              text-anchor="middle"
              textLength="40"
              x="72"
              y="87"
            >
              5
            </text>
          </svg>
          
        </div>
        <h3>
          Raises, as young
           
          
        </h3>
        <p>
          By 
          Mike D
        </p>
      </a>
    </div>
    <p
      class="emotion-16"
    >
      For questions and discussion, join the 
      <a
        href="https://groups.google.com/forum/#!forum/crosshare"
        rel="noopener noreferrer"
        target="_blank"
      >
        Google Group
      </a>
      . Follow us on twitter 
      <a
        href="https://twitter.com/crosshareapp"
        rel="noopener noreferrer"
        target="_blank"
      >
        @crosshareapp
      </a>
      .
    </p>
  </div>
</div>
`;
